### PR TITLE
fix(client): Fix URL parsing in the CLI

### PR DESCRIPTION
**Pull request by alco**
This is just to demonstrate the direction in which URL parsing can be improved in the CLI.

First, more test coverage is needed for some not-too-uncommon edge cases.

Second, more uniformunity is needed in the parsing code. Currently, in `src/cli/utils.ts` alone the database URL is parsed using a regexp, the service URL is parsed using the deprecated `url.parse()` method. All of this can be replaced with the built-in `URL` class and its properties/method. That class can also be used for building URLs, instead of ad-hoc string concatenation that doesn't escape URL components.

### CODE REVIEW

**Comment by msfstef** [comment]
@alco thank you for the fix! I've applied the suggested changes and added a couple of tests for some of them. I've also refactored the utilities slightly under the existing `util` folder to avoid having both a `util` folder and a root level `utils.js` file.

You've mentioned some not-too-uncommon edge cases, if you have some in mind could you share them so we can add tests for those as well?

**Review by msfstef — APPROVED** [review]

**Comment by alco** [comment]
> You've mentioned some not-too-uncommon edge cases, if you have some in mind could you share them so we can add tests for those as well?

Check out the [doc tests](https://github.com/electric-sql/electric/blob/d406e85025ef4229ac1890843a1be5c3b4603b57/components/electric/lib/electric/config.ex#L148-L238) for database URL parsing in the server code. It'd be great to not have to duplicate those checks and error messages, but I don't think there's a way around that. The client has to do its own parsing to extract connection options from the URL. It should also have descriptive errors for invalid URLs.

**Comment by alco** [comment]
By the way, thanks for picking this up after my initial draft commits, @msfstef !

**Comment by msfstef** [comment]
@alco thank _you_ for getting this PR opened :) I've added some of the test cases in the TS client as well, some are covered by the `URL` API (e.g. missing host is an invalid URL), and some I've adjusted (e.g. IPv6 gets parsed with brackets, but I think that's safe to use that way) - added descriptive errors where appropriate

**Review by kevin-dp — APPROVED** [review]
Good work! I left some minor comments.

File: clients/typescript/src/cli/util/parse.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+export function parsePgProxyPort(str: string | number): {
+  http: boolean
+  port: number
+} {
+  if (typeof str === 'number') {
+    return { http: false, port: str }
+  } else if (str.includes(':')) {
+    const [prefix, port] = str.split(':')
```
[the remark this answers is not in the document]
Comment (kevin-dp) [reply clients/typescript/src/cli/util/parse.ts:79]: There are no unit tests for this function.

File: clients/typescript/src/cli/util/parse.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import path from 'path'
+import { InvalidArgumentError } from 'commander'
+import { appRoot } from './paths'
+
+/**
+ * Get the name of the current project.
+ */
+export function getAppName(): string | undefined {
+  const packageJsonPath = path.join(appRoot, 'package.json')
+  return JSON.parse(fs.readFileSync(packageJsonPath, 'utf8')).name
```
[the remark this answers is not in the document]
Comment (msfstef) [reply clients/typescript/src/cli/util/parse.ts:11]: apparently `require` does some caching under the hood, and requires a linter suppression to use which is annoying, so I'm going to stick with the `fs` and parse approach for now - if it ain't broke haha

File: clients/typescript/src/cli/util/parse.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+export function extractDatabaseURL(url: string): {
+  user: string
+  password: string
+  host: string
+  port: number | null
+  dbName: string
+} {
+  const parsed = new URL(url)
+  if (!(parsed.protocol == 'postgres:' || parsed.protocol == 'postgresql:')) {
```
[the remark this answers is not in the document]
Comment (msfstef) [reply clients/typescript/src/cli/util/parse.ts:40]: huh this really should be a lint rule (the `eqeqeq` one) - I'll add it on this PR and fix it in several places cause it seems we're using double equality often https://github.com/electric-sql/electric/pull/1313

File: clients/typescript/src/cli/util/parse.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (msfstef) [reply clients/typescript/src/cli/util/parse.ts:94]: We actually don't document this clearly but if you specify the `PG_PROXY_PORT` as `http` or `http:{PORT}` then the sync service will use the http proxy tunnel (see https://electric-sql.com/docs/api/cli#proxy-tunnel)

I've modified the type signature for the function and added a few unit tests, didn't cover all invalid scenarios as we can rely on typescript's checking

File: clients/typescript/src/cli/util/parse.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (msfstef) [reply clients/typescript/src/cli/util/parse.ts]: See this commit https://github.com/electric-sql/electric/pull/1313/commits/80024eebcf5bdea074b8e5463a2a7c35e8fd4886 I've had to add the rule to eslint but some cases required some thought as double equality was being abused

**Comment by alco** [comment]
@msfstef Feel free to merge this when it's ready.
